File: packages/bruno-ios/ios/App/Plugins/BrunoNative/Sources/BrunoNative/PathModel.swift
import Foundation

/// Single rebasing authority between device-absolute filesystem paths and the
/// container-relative `@documents` tokens that cross the bridge to JS.
///
/// The app sandbox container UUID changes on every reinstall, so no absolute
/// path may ever be persisted. Everything outbound is tokenized; everything
/// inbound is resolved against the current Documents root before touching the
/// filesystem.
enum PathModel {
    /// Logical root token literal.
    static let token = "@documents"

    /// Overridable Documents root. Defaults to the FileManager Documents
    /// directory (canonicalized). Tests inject a temporary directory because the
    /// test host's Documents directory differs from the app sandbox.
    static var documentsRootOverride: URL?

    static var documentsRoot: URL {
        if let override = documentsRootOverride {
            return override.resolvingSymlinksInPath()
        }
        let url = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        return url.resolvingSymlinksInPath()
    }

    static func isToken(_ path: String) -> Bool {
        return path == token || path.hasPrefix(token + "/")
    }

    /// Collapses duplicate slashes, `.` segments and trailing slashes. Returns
    /// the bare root token unchanged.
    static func normalizeToken(_ value: String) -> String {
        guard isToken(value) else {
            return value
        }
        let relative = String(value.dropFirst(token.count))
        let segments = relative
            .split(separator: "/", omittingEmptySubsequences: true)
            .filter { $0 != "." }
        if segments.isEmpty {
            return token
        }
        return token + "/" + segments.joined(separator: "/")
    }

    /// Converts a device-absolute path under the Documents root into a token.
    /// A path outside the root is returned unchanged (guard, not a feature).
    static func tokenize(_ absolutePath: String) -> String {
        let rootPath = documentsRoot.path
        let normalizedAbsolute = URL(fileURLWithPath: absolutePath).resolvingSymlinksInPath().path

        if normalizedAbsolute == rootPath {
            return token
        }
        let rootPrefix = rootPath.hasSuffix("/") ? rootPath : rootPath + "/"
        guard normalizedAbsolute.hasPrefix(rootPrefix) else {
            return absolutePath
        }
        let relative = String(normalizedAbsolute.dropFirst(rootPrefix.count))
        if relative.isEmpty {
            return token
        }
        return token + "/" + relative
    }

    /// Converts a token into a device-absolute path under the current Documents
    /// root. A value that is already absolute and not a token is passed through.
    static func resolve(_ tokenPath: String) -> String {
        guard isToken(tokenPath) else {
            return tokenPath
        }
        let normalized = normalizeToken(tokenPath)
        if normalized == token {
            return documentsRoot.path
        }
        let relative = String(normalized.dropFirst(token.count + 1))
        return documentsRoot.appendingPathComponent(relative).path
    }

    /// Deep-walks an argument value resolving tokens to absolute paths. Used for
    /// structured inbound args (the snapshot blob, nested objects).
    static func rebaseInbound(_ value: Any) -> Any {
        return walkSnapshot(value, transform: resolve)
    }

    /// Deep-walks a value tokenizing absolute paths in known path-bearing
    /// fields. Used for return values and event payloads.
    static func rebaseOutbound(_ value: Any) -> Any {
        return walkSnapshot(value, transform: tokenize)
    }

    // MARK: - Snapshot walk

    /// The absolute-path fields the snapshot walk rebases. A schema-aware
    /// traversal (not blind) so only known path-bearing fields are rewritten.
    private static let workspacePathKeys: Set<String> = [
        "pathname",
        "lastActiveCollectionPathname"
    ]
    private static let collectionPathKeys: Set<String> = [
        "pathname",
        "workspacePathname",
        "environmentPath"
    ]
    private static let tabPathKeys: Set<String> = [
        "pathname"
    ]

    /// Schema-aware traversal over the 9 snapshot path fields. The `transform`
    /// is either `resolve` (read) or `tokenize` (write).
    static func walkSnapshot(_ value: Any, transform: (String) -> String) -> Any {
        guard var snapshot = value as? [String: Any] else {
            return value
        }

        if let activeWorkspacePath = snapshot["activeWorkspacePath"] as? String {
            snapshot["activeWorkspacePath"] = transform(activeWorkspacePath)
        }

        if let workspaces = snapshot["workspaces"] as? [[String: Any]] {
            snapshot["workspaces"] = workspaces.map { rebaseWorkspace($0, transform: transform) }
        }

        if let collections = snapshot["collections"] as? [[String: Any]] {
            snapshot["collections"] = collections.map { rebaseCollection($0, transform: transform) }
        }

        return snapshot
    }

    private static func rebaseWorkspace(_ value: [String: Any], transform: (String) -> String) -> [String: Any] {
        var workspace = value
        for key in workspacePathKeys {
            if let path = workspace[key] as? String {
                workspace[key] = transform(path)
            }
        }
        if let collections = workspace["collections"] as? [String] {
            workspace["collections"] = collections.map(transform)
        }
        return workspace
    }

    private static func rebaseCollection(_ value: [String: Any], transform: (String) -> String) -> [String: Any] {
        var collection = value
        for key in collectionPathKeys {
            if let path = collection[key] as? String {
                collection[key] = transform(path)
            }
        }
        if var environment = collection["environment"] as? [String: Any] {
            if let collectionPath = environment["collection"] as? String {
                environment["collection"] = transform(collectionPath)
            }
            collection["environment"] = environment
        }
        if let tabs = collection["tabs"] as? [[String: Any]] {
            collection["tabs"] = tabs.map { tab -> [String: Any] in
                var entry = tab
                for key in tabPathKeys {
                    if let path = entry[key] as? String {
                        entry[key] = transform(path)
                    }
                }
                return entry
            }
        }
        return collection
    }
}

File: packages/bruno-ios/src/path-model.ts
/**
 * Container-relative PathModel — JS mirror of PathModel.swift.
 *
 * The iOS app sandbox Documents directory changes its absolute location on every
 * (re)install (the container UUID rotates), so no device-absolute path may ever be
 * persisted in Redux/YAML/snapshots. Instead the app speaks in '@documents' tokens:
 * a stable, install-independent logical path. Swift is the single rebasing authority
 * at the bridge boundary; this module is the testable mirror used for the Jest
 * equivalence suite, normalizeToken comparisons during snapshot hydration, and a
 * defensive rebaseSnapshotForWrite pass before a snapshot is handed to Swift.
 *
 * Every function is pure: the Documents root is always supplied as a parameter; this
 * module never touches FileManager or any native API.
 */

/** Logical root token literal. The bare root is '@documents' (no trailing slash). */
export const TOKEN = '@documents';

/** Strips a single trailing slash (never the bare root token). */
const stripTrailingSlash = (value: string): string => {
  if (value.length > 1 && value.endsWith('/')) {
    return value.replace(/\/+$/, '');
  }
  return value;
};

/** True when the path starts with the '@documents' token. */
export const isToken = (path: string): boolean => {
  return path === TOKEN || path.startsWith(`${TOKEN}/`);
};

/**
 * Collapses '//', './' and trailing slashes so two tokens that name the same
 * location compare equal as exact strings. Leaves a non-token path unchanged.
 */
export const normalizeToken = (token: string): string => {
  if (!isToken(token)) {
    return token;
  }
  const rest = token.slice(TOKEN.length);
  const segments = rest.split('/').filter((segment) => segment !== '' && segment !== '.');
  if (segments.length === 0) {
    return TOKEN;
  }
  return `${TOKEN}/${segments.join('/')}`;
};

/**
 * Maps an absolute path under documentsRoot to its '@documents/<relative>' token.
 * The bare root maps to '@documents'. A path outside documentsRoot is returned
 * unchanged (a guard, not a feature — v1 only ever operates inside Documents).
 */
export const tokenize = (absPath: string, documentsRoot: string): string => {
  const root = stripTrailingSlash(documentsRoot);
  if (absPath === root) {
    return TOKEN;
  }
  if (absPath.startsWith(`${root}/`)) {
    const relative = absPath.slice(root.length + 1);
    return normalizeToken(`${TOKEN}/${relative}`);
  }
  return absPath;
};

/**
 * Maps a '@documents/<relative>' token to an absolute path under documentsRoot.
 * The bare root token maps to documentsRoot. A non-token path is returned
 * unchanged (passthrough for already-absolute values).
 */
export const resolve = (tokenPath: string, documentsRoot: string): string => {
  if (!isToken(tokenPath)) {
    return tokenPath;
  }
  const root = stripTrailingSlash(documentsRoot);
  const normalized = normalizeToken(tokenPath);
  if (normalized === TOKEN) {
    return root;
  }
  const relative = normalized.slice(TOKEN.length + 1);
  return `${root}/${relative}`;
};

/**
 * The absolute-path fields a snapshot walk must rebase. Schema-aware (not blind)
 * so only known path-bearing fields are rewritten — names/values that merely look
 * path-like are left untouched.
 */
const rebaseSnapshot = (
  blob: any,
  mapPath: (value: string) => string
): any => {
  if (!blob || typeof blob !== 'object') {
    return blob;
  }

  const mapMaybe = (value: any): any => (typeof value === 'string' ? mapPath(value) : value);

  const next: any = { ...blob };

  if ('activeWorkspacePath' in next) {
    next.activeWorkspacePath = mapMaybe(next.activeWorkspacePath);
  }

  if (Array.isArray(next.workspaces)) {
    next.workspaces = next.workspaces.map((workspace: any) => {
      if (!workspace || typeof workspace !== 'object') {
        return workspace;
      }
      const nextWorkspace: any = { ...workspace };
      if ('pathname' in nextWorkspace) {
        nextWorkspace.pathname = mapMaybe(nextWorkspace.pathname);
      }
      if ('lastActiveCollectionPathname' in nextWorkspace) {
        nextWorkspace.lastActiveCollectionPathname = mapMaybe(nextWorkspace.lastActiveCollectionPathname);
      }
      if (Array.isArray(nextWorkspace.collections)) {
        nextWorkspace.collections = nextWorkspace.collections.map(mapMaybe);
      }
      return nextWorkspace;
    });
  }

  if (Array.isArray(next.collections)) {
    next.collections = next.collections.map((collection: any) => {
      if (!collection || typeof collection !== 'object') {
        return collection;
      }
      const nextCollection: any = { ...collection };
      if ('pathname' in nextCollection) {
        nextCollection.pathname = mapMaybe(nextCollection.pathname);
      }
      if ('workspacePathname' in nextCollection) {
        nextCollection.workspacePathname = mapMaybe(nextCollection.workspacePathname);
      }
      if ('environmentPath' in nextCollection) {
        nextCollection.environmentPath = mapMaybe(nextCollection.environmentPath);
      }
      if (nextCollection.environment && typeof nextCollection.environment === 'object') {
        const nextEnvironment: any = { ...nextCollection.environment };
        if ('collection' in nextEnvironment) {
          nextEnvironment.collection = mapMaybe(nextEnvironment.collection);
        }
        nextCollection.environment = nextEnvironment;
      }
      if (Array.isArray(nextCollection.tabs)) {
        nextCollection.tabs = nextCollection.tabs.map((tab: any) => {
          if (!tab || typeof tab !== 'object') {
            return tab;
          }
          const nextTab: any = { ...tab };
          if ('pathname' in nextTab) {
            nextTab.pathname = mapMaybe(nextTab.pathname);
          }
          return nextTab;
        });
      }
      return nextCollection;
    });
  }

  return next;
};

/** Walks the 9 snapshot path fields, mapping tokens to absolute paths. */
export const rebaseSnapshotForRead = (blob: any, documentsRoot: string): any => {
  return rebaseSnapshot(blob, (value) => resolve(value, documentsRoot));
};

/** Walks the 9 snapshot path fields, mapping absolute paths to tokens. */
export const rebaseSnapshotForWrite = (blob: any, documentsRoot: string): any => {
  return rebaseSnapshot(blob, (value) => tokenize(value, documentsRoot));
};
